nav2_local_planner_interface/include/nav2_local_planner_interface/local_planner_base.hpp

```diff
@@ -0,0 +1,118 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2017, Locus Robotics
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef NAV2_LOCAL_PLANNER_INTERFACE__LOCAL_PLANNER_BASE_HPP_
```

Comment:
How much design thought when into the changes for the local planner abstract interface? it looks like you copied david lus nearly -- I thought we wanted the goal checker in the BT? we have the node for it. 

lets not use `nav_2d_msgs` its extremely limiting. David chose to do that to my understanding because Locus doesn't need it and he doesn't want to support non-2D modes, per his talk in 2018. One of the core bullets on the pipeline for nav2 is to support more robots like ackermann or non-flat grounds which this restricts from the get go. Even if we still use 2D information for now, if we lock in this local planning API to only use 2D defined messages, it'll break everyone down the line if/when we change

Reply:
> How much design thought when into the changes for the local planner abstract interface? it looks like you copied david lus nearly -- I thought we wanted the goal checker in the BT? we have the node for it.

Hmm. Didn't think about this. Yes, it mostly structured like David Lu's. I agree that there could be a better place for the goal checker.

> ets not use `nav_2d_msgs` its extremely limiting. David chose to do that to my understanding because Locus doesn't need it and he doesn't want to support non-2D modes, per his talk in 2018. One of the core bullets on the pipeline for nav2 is to support more robots like ackermann or non-flat grounds which this restricts from the get go. Even if we still use 2D information for now, if we lock in this local planning API to only use 2D defined messages, it'll break everyone down the line if/when we change

True. I just copied those from the what is currently in the stack in '/nav2_dwb_controller/nav_2d_msgs/'.
I was thinking of addressing that in a different PR.

Reply:
> I thought we wanted the goal checker in the BT?

It's something we talked about, but it is a more radical change. To stop, we'd be adding some extra latency to do the network communication and stuff to cancel the outstanding action.

Also, we'd be bringing in a significantly more complex node into the BT tree, which you were indicating you wanted to avoid.

If we do want to bring this into the BT, I'm not sure we should do it quite yet. Both the action cancelling code in our stack and the networking in ROS are flaky at the moment.

Reply:
So I’m not saying we Need to change anything about the navcore2 headers, but I think there should be a bit of discussion around design to make sure theres not other things we want to add (or delete, rename, different message types). I havent put enough thought into this to have a concrete example. 

100% wipe nav 2d messages from the API, I want full 3D support should someone want to do it (me). Even if DWB immediately on call reverts to nav 2d msgs for its own use, it doesnt shoot us in the foot come ackermann time.

Reply:
@crdelsey I dont know that I do or dont want it in the BT. I just saw the BT goal condition and I assumed that’s where you wanted to go with this, which I would support. 

If the behavior tree is the arbiter of information, calling a planner, giving that path to a controller, calling a controller. It wouldn’t break that model to also call a goal checker, give that goal status info to relevant parties (and cancel the action, or pass that goal status to the actions for them to decide their next action)

Reply:
Something we might want to add to the interface is a `getStatusCode()` type method with an enum of statuses like failure, types of failure, running, idle, etc and then relay that information in the action server back to the behavior tree as context